src/serial.rs

```diff
@@ -1398,388 +1399,474 @@ pub struct Tx<USART> {
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479"
+)))]
+use crate::pac::usart1 as uart_base;
 
-                fn bwrite_all(&mut self, bytes: &[u8]) -> Result<(), Self::Error> {
-                    let mut tx: Tx<$USARTX> = Tx {
-                        _usart: PhantomData,
-                    };
-                    tx.bwrite_all(bytes)
-                }
+mod private {
+    pub trait Sealed {}
+}
 
-                fn bflush(&mut self) -> Result<(), Self::Error> {
-                    let mut tx: Tx<$USARTX> = Tx {
-                        _usart: PhantomData,
-                    };
-                    tx.bflush()
-                }
-            }
-        )+
-    }
+// Implemented by all USART instances
+pub trait Instance: private::Sealed {
+    #[doc(hidden)]
+    fn ptr() -> *const uart_base::RegisterBlock;
+    #[doc(hidden)]
+    fn pclk_freq(clocks: &Clocks) -> u32;
+    #[doc(hidden)]
+    unsafe fn enable_clock(rcc: &crate::stm32::rcc::RegisterBlock);
+    #[doc(hidden)]
+    fn stopbits(bits: config::StopBits) -> uart_base::cr2::STOP_A;
 }
 
 macro_rules! halUsart {
     ($(
         $USARTX:ident: ($usartX:ident, $apbXenr:ident, $rcc_bit:expr, $usartXen:ident, $pclkX:ident),
     )+) => {
         $(
-        impl<PINS> Serial<$USARTX, PINS> {
-            fn config_stop(self, config: config::Config) -> Self {
-                use crate::stm32::usart1::cr2::STOP_A;
-                use self::config::*;
-
-                self.usart.cr2.write(|w| {
-                    w.stop().variant(match config.stopbits {
-                        StopBits::STOP0P5 => STOP_A::STOP0P5,
-                        StopBits::STOP1 => STOP_A::STOP1,
-                        StopBits::STOP1P5 => STOP_A::STOP1P5,
-                        StopBits::STOP2 => STOP_A::STOP2,
-                    })
-                });
-                self
+            impl private::Sealed for $USARTX {}
+            impl Instance for $USARTX {
+                fn ptr() -> *const uart_base::RegisterBlock {
+                    $USARTX::ptr() as *const _
+                }
+
+                fn pclk_freq(clocks: &Clocks) -> u32 {
+                    clocks.$pclkX().0
+                }
+
+                unsafe fn enable_clock(rcc: &crate::stm32::rcc::RegisterBlock) {
+                    use crate::bb;
+
+                    bb::set(&rcc.$apbXenr, $rcc_bit);
+                }
+
+                fn stopbits(bits: config::StopBits) -> uart_base::cr2::STOP_A {
+                    use crate::stm32::usart1::cr2::STOP_A;
+                    use config::StopBits;
+
+                    unsafe {
+                        mem::transmute(match bits {
+                            StopBits::STOP0P5 => STOP_A::STOP0P5,
+                            StopBits::STOP1 => STOP_A::STOP1,
+                            StopBits::STOP1P5 => STOP_A::STOP1P5,
+                            StopBits::STOP2 => STOP_A::STOP2,
+                        })
```

Comment:
`uart_base` is sometimes `uart4`, which only has two variants for this enum, so this `transmute` is unsound.